File: lib/src/model/camera_position.dart
part of '../../flutter_kakao_maps.dart';

class CameraPosition {
  final double height;
  final double rotationAngle;
  final double tiltAngle;
  final int zoomLevel;
  final KakaoMapPoint position;
  KakaoMapGestureType? gestureType;

  CameraPosition({
    required this.height,
    required this.rotationAngle,
    required this.tiltAngle,
    required this.zoomLevel,
    required this.position,
    this.gestureType,
  });

  factory CameraPosition.fromJson(Map<String, dynamic> json) {
    final rotationAngle = json['rotationAngle'];
    final tiltAngle = json['tiltAngle'];
    final height = json['height'];
    final gestureType = json['gestureType'];

    return CameraPosition(
      gestureType: gestureType is String
          ? gestureType.toStatus()
          : KakaoMapGestureType.unknown,
      height: height is int ? height.toDouble() : height,
      rotationAngle:
          rotationAngle is int ? rotationAngle.toDouble() : rotationAngle,
      tiltAngle: tiltAngle is int ? tiltAngle.toDouble() : tiltAngle,
      zoomLevel: json['zoomLevel'],
      position: KakaoMapPoint.fromJson(json['position']),
    );
  }
}

File: android/src/main/kotlin/com/shareinvest/flutter_kakao_maps/model/CameraPosition.kt
package com.shareinvest.flutter_kakao_maps.model

import com.kakao.vectormap.LatLng
import org.json.JSONObject

data class CameraPosition(
    var height: Double,
    var position: LatLng,
    var rotationAngle: Double,
    var tiltAngle: Double,
    var zoomLevel: Int,
)

fun JSONObject.toCameraPosition(): CameraPosition {
    val position = this.getJSONObject("position")

    return CameraPosition(
        height = this.getDouble("height"),
        rotationAngle = this.getDouble("rotationAngle"),
        tiltAngle = this.getDouble("tiltAngle"),
        zoomLevel = this.getInt("zoomLevel"),
        position = LatLng.from(
            position.getDouble("latitude"),
            position.getDouble("longitude")
        ),
    )
}